pipit/util/test_generator.py

```diff
@@ -0,0 +1,366 @@
+            [
+                textwrap.indent(str(run_time) + ": " + str(child), "\t")
+                for run_time, child in sorted(self.children.items())
+            ]
+        )
+
+    def mpi_attributes(self):
+        """
+        Returns the Attributes dictionary for this node, including
+        (if it is an MPI event) receiver/sender and msg_length.
+        """
+        if not self.is_mpi:
+            return {}
+        attr = {"msg_length": self.mpi_volume}
+        if self.mpi_type == "MpiSend":
+            attr["receiver"] = self.mpi_tgt
+        else:
+            attr["sender"] = self.mpi_tgt
+        return attr
+
+    def to_events(self, begin_time, process, data):
+        """
+        Returns event data for this tree, with time starting at begin_time.
+        data is an array that is built up and then converted to a DataFrame
+        once the entire tree has been processed.
+        """
+        data.append(
+            [
+                begin_time,
+                "Enter",
+                self.name + "()",
+                process,
+                {},
+                self.inc_time,
+                self.exc_time,
+            ]
+        )
+        if self.is_mpi:
+            data.append(
+                [
+                    begin_time + self.mpi_time,
+                    "Instant",
+                    self.mpi_type,
+                    process,
+                    self.mpi_attributes(),
+                    float("nan"),
+                    float("nan"),
+                ]
+            )
+        # total_time accumulates durations of already processed children
+        total_time = begin_time
+        for run_time, child in sorted(self.children.items()):
+            # children will add their own lines to data
+            child.to_events(total_time + run_time, process, data)
+            total_time += child.inc_time
+
+        # time.inc and time.exc are both NaN for Leave events
+        data.append(
+            [
+                begin_time + self.inc_time,
+                "Leave",
+                self.name + "()",
+                process,
+                {},
+                float("nan"),
+                float("nan"),
+            ]
+        )
+
+    def tweak_tree(self):
+        """
+        Adds small exclusive time perturbations to
+        function length to generate a "similar" tree, for other processes
+        to use, that is not identical to the original.
+        """
+        exc_time = self.exc_time
+        factor = (0.7) + random.random() * 0.6
+        exc_time_new = exc_time * factor
+        # preserve inclusive time relations
+        self.grow_inc_time(exc_time_new - exc_time)
+        self.exc_time = exc_time_new
+        for run_time, child in self.children.items():
+            child.tweak_tree()
+        # also scale back child run times
+        self.children = {
+            run_time * factor: child for run_time, child in self.children.items()
```

Comment:
I think I see what you're trying to do here, but I'm not sure it's foolproof. I might be misunderstanding what exactly `run_time` is, but I feel like even with scaling the runtimes, there are cases where some child can go outside of the parent's time interval after this transformation.